Optimization docs

akihironitta opened the pull request:
## What does this PR do?
Fixes #<issue_number> Follow-up of https://github.com/PyTorchLightning/pytorch-lightning/pull/6825#issuecomment-815692273. Also requested by the community in #5780.

---

### Updated docs links (23803e3)
**Optimization page**
Current: https://pytorch-lightning.readthedocs.io/en/latest/common/optimizers.html
Updated: https://108739-178626720-gh.circle-artifacts.com/0/html/common/optimizers.html

**LightningModule page**
Current: https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html
Updated:  https://108739-178626720-gh.circle-artifacts.com/0/html/common/lightning_module.html

---

### Motivation
In my opinion, [the current docs](https://pytorch-lightning.readthedocs.io/en/latest/common/optimizers.html) about optimization (especially about manual optimization) are somewhat messy. There are quite a number of tip/note/warning which are randomly distributed. Also, I found a few outdated/wrong examples in the docs, so I'll try to address them here.

---

### Description of the changes
- Restructure the optimization page
- Verify that examples are correct
- Other small fixes/improvements of typo, indent, ...

---

## Before submitting
- [n/a] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [n/a] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [n/a] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

<!-- For CHANGELOG separate each item in the unreleased section by a blank line to reduce collisions -->

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

akihironitta commented on docs/source/common/optimizers.rst:
```diff
@@ -3,238 +3,322 @@
+       # do forward, backward, and optimization
+       ...
 
-        def closure():
-            # Only zero_grad on the first batch to accumulate gradients
-            is_first_batch_to_accumulate = batch_idx % 2 == 0
-            if is_first_batch_to_accumulate:
-                opt.zero_grad()
```
When using `LBFGS`-like optimizers which require a closure function to reevaluate the model, the closure always has to run `zero_grad` to make them work properly, so "gradient accumulation + closure" is not supported in PL as well as PyTorch, I think.

akihironitta commented on docs/source/common/optimizers.rst:
```diff
@@ -390,100 +487,65 @@ Lightning will call each optimizer sequentially:
-
-----------
-
-Using the closure functions for optimization
```
This section is completely removed because
1. from 1.2.2, this isn't needed even when using LBFGS-like optimizers
2. It doesn't include `zero_grad` in the closure function, so this example is wrong anyway.

akihironitta commented:
current: https://pytorch-lightning.readthedocs.io/en/latest/common/optimizers.html#learning-rate-scheduling
updated: https://107599-178626720-gh.circle-artifacts.com/0/html/common/optimizers.html

akihironitta commented:
current: https://pytorch-lightning.readthedocs.io/en/latest/common/optimizers.html#learning-rate-scheduling
updated: https://108108-178626720-gh.circle-artifacts.com/0/html/common/optimizers.html

Borda reviewed (approved):
lgtm 🐰

awaelchli commented on docs/source/common/optimizers.rst:
```diff
@@ -3,27 +3,38 @@
 
-* Set the ``automatic_optimization`` property to ``False`` in your ``LightningModule`` ``__init__`` function
-* Use ``self.manual_backward(loss)`` instead of ``loss.backward()``.
+  * ``optimizer.zero_grad()`` to clear the gradients from the previous training step
+  * ``self.manual_backward(loss)`` instead of ``loss.backward()``
+  * ``optimizer.step()`` to update your model parameters
```
```suggestion
  * ``self.optimizers()`` to access your optimizers (one or multiple)
  * ``optimizer.zero_grad()`` to clear the gradients from the previous training step
  * ``self.manual_backward(loss)`` instead of ``loss.backward()``
  * ``optimizer.step()`` to update your model parameters
```

awaelchli reviewed (approved):
love this doc update! thanks

carmocca commented on docs/source/common/optimizers.rst:
```diff
@@ -3,27 +3,38 @@
+
+To manually optimize, do the following:
+
+* Set the ``automatic_optimization`` property to ``False`` in your ``LightningModule`` ``__init__`` function.
```
```suggestion
* Set the ``self.automatic_optimization = False`` in your ``LightningModule``'s ``__init__``.
```

carmocca commented on docs/source/common/optimizers.rst:
```diff
@@ -32,25 +43,40 @@ to manually manage the optimization process. To do so, do the following:
-.. note:: This is only recommended for experts who need ultimate flexibility. Lightning will handle only precision and accelerators logic. The users are left with ``optimizer.zero_grad()``, gradient accumulation, model toggling, etc..
+.. warning::
+   Before 1.2, ``optimizer.step()`` was calling ``optimizer.zero_grad()`` internally.
+   From 1.2, it is left to the users expertise.
```
```suggestion
   From 1.2, it is left to the user's expertise.
```

carmocca commented on docs/source/common/optimizers.rst:
```diff
@@ -159,32 +164,98 @@ Here is the same example as above using a ``closure``.
+Learning rate scheduling [manual]
+---------------------------------
+You can call ``lr_scheduler.step()`` at arbitrary intervals.
+Use ``self.lr_schedulers()`` in LightningModule to access your learning rate schedulers defined
```
```suggestion
Use ``self.lr_schedulers()`` in  your :class:`~pytorch_lightning.LightningModule` to access any learning rate schedulers defined
```

carmocca commented on docs/source/common/optimizers.rst:
```diff
@@ -159,32 +164,98 @@ Here is the same example as above using a ``closure``.
+---------------------------------
+You can call ``lr_scheduler.step()`` at arbitrary intervals.
+Use ``self.lr_schedulers()`` in LightningModule to access your learning rate schedulers defined
+in your ``LightningModule.configure_optimizers()``.
```
```suggestion
in your :meth:`~pytorch_lightning.LightningModule.configure_optimizers`.
```

carmocca commented on docs/source/common/optimizers.rst:
```diff
@@ -159,32 +164,98 @@ Here is the same example as above using a ``closure``.
+
+.. warning::
+   * Before 1.3, Lightning automatically calls ``lr_scheduler.step()`` in both automatic and manual optimization. From
+     1.3, ``lr_scheduler.step()`` is disabled in manual optimization so that you can call it at arbitrary intervals.
```
```suggestion
     1.3, ``lr_scheduler.step()`` is now for the user to call at arbitrary intervals.
```

carmocca commented on docs/source/common/optimizers.rst:
```diff
@@ -159,32 +164,98 @@ Here is the same example as above using a ``closure``.
+in your ``LightningModule.configure_optimizers()``.
+
+.. warning::
+   * Before 1.3, Lightning automatically calls ``lr_scheduler.step()`` in both automatic and manual optimization. From
```
```suggestion
   * Before 1.3, Lightning automatically called ``lr_scheduler.step()`` in both automatic and manual optimization. From
```

carmocca commented on docs/source/common/optimizers.rst:
```diff
@@ -159,32 +164,98 @@ Here is the same example as above using a ``closure``.
+   * Before 1.3, Lightning automatically calls ``lr_scheduler.step()`` in both automatic and manual optimization. From
+     1.3, ``lr_scheduler.step()`` is disabled in manual optimization so that you can call it at arbitrary intervals.
+   * Note that the lr_dict keys, such as ``"step"`` and ``""interval"``, will be ignored even if they are provided in
+     your ``configure_optimizers()`` in manual optimization.
```
```suggestion
     your ``configure_optimizers()`` during manual optimization.
```

carmocca commented on docs/source/common/optimizers.rst:
```diff
@@ -269,221 +372,209 @@ In the case of multiple optimizers, Lightning does the following:
-        # on_after_backward callback
-        pl_module.on_after_backward(result.training_step_output, batch_idx, result.loss)
-
-        return result
-
-    # This default `second_order_closure` function can be enabled by passing it directly into the `optimizer.step`
-    def optimizer_step(self, current_epoch, batch_nb, optimizer, optimizer_idx, second_order_closure, on_tpu=False, using_native_amp=False, using_lbfgs=False):
-        # update params
-        optimizer.step(second_order_closure)
+    def optimizer_step(
+        self, epoch, batch_idx, optimizer, optimizer_idx, optimizer_closure,
+        on_tpu=False, using_native_amp=False, using_lbfgs=False,
+    ):
+        # `optimizer` is a `LightningOptimizer` wrapping the optimizer.
+        # To access it, do as follow:
+        optimizer = optimizer.optimizer
+
+        # run step. However, it won't work on TPU, AMP, etc...
+        optimizer.step(closure=optimizer_closure)
```
You should merge this with the note in

https://github.com/PyTorchLightning/pytorch-lightning/pull/6907/files#diff-f91783b13a45bd611ed87185b906c97af54c613882eeee65adcf8ec89b1723c2R60-R65

akihironitta replied:
Updated the note there. Mind having a look again?

carmocca commented on docs/source/common/optimizers.rst:
```diff
@@ -269,221 +372,209 @@ In the case of multiple optimizers, Lightning does the following:
 .. note::
-    Metrics can be made availble to condition on by simply logging it using ``self.log('metric_to_track', metric_val)``
-    in your lightning module.
+   Metrics can be made available to condition on by simply logging it using ``self.log('metric_to_track', metric_val)``
```
```suggestion
   Metrics can be made available to monitor by simply logging it using ``self.log('metric_to_track', metric_val)``
```

carmocca reviewed (commented):
Awesome work! We really needed this

carmocca commented on docs/source/common/optimizers.rst:
```diff
@@ -32,25 +44,40 @@ to manually manage the optimization process. To do so, do the following:
 
-.. warning:: Before 1.2, ``optimzer.step`` was calling ``optimizer.zero_grad()`` internally. From 1.2, it is left to the users expertise.
+.. tip::
+   * ``self.optimizers()`` will return :class:`~pytorch_lightning.core.optimizer.LightningOptimizer` objects. You can
+     access your own optimizer with ``optimizer.optimizer``. However, if you use your own optimizer to perform a step,
+     Lightning won't be able to support accelerators and precision for you.
+   * Be careful where you call ``optimizer.zero_grad()``, or your model won't converge.
+     It is good practice to call ``optimizer.zero_grad()`` before ``self.manual_backward(loss)``.
```
Remove this now that we have it at the bottom?

```suggestion
   Be careful where you call ``optimizer.zero_grad()``, or your model won't converge.
   It is good practice to call ``optimizer.zero_grad()`` before ``self.manual_backward(loss)``.
```

akihironitta replied:
I would leave it in both manual and automatic optimization sections as one might only look at a section of their interest, but the decision is yours. Should I apply your suggestion here?

akihironitta commented on pytorch_lightning/core/lightning.py:
```diff
@@ -1359,22 +1352,18 @@ def optimizer_step(self, epoch, batch_idx, optimizer, optimizer_idx,
+                # update discriminator opt every 2 steps
                 if optimizer_idx == 1:
-                    if batch_idx % 4 == 0 :
+                    if (batch_idx + 1) % 2 == 0 :
```
Just for the note, either optimizer needs to update its parameters every step. Otherwise, half of the training dataset will be ignored in this case...

akihironitta commented on pytorch_lightning/core/lightning.py:
```diff
@@ -1210,23 +1202,24 @@ def configure_optimizers(self):
-        .. tip:: In manual mode we still automatically clip grads if Trainer(gradient_clip_val=x) is set
+        See :ref:`manual optimization<common/optimizers:Manual optimization>` for more examples.
 
-        .. tip:: In manual mode we still automatically accumulate grad over batches if
```
As pointed out by @rubencart, this tip is outdated.

akihironitta commented:
Here are the links to the pages that are mainly changed.

**Optimization page**
Current: https://pytorch-lightning.readthedocs.io/en/latest/common/optimizers.html
Updated: https://108701-178626720-gh.circle-artifacts.com/0/html/common/optimizers.html

**LightningModule page**
Current: https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html
Updated:  https://108701-178626720-gh.circle-artifacts.com/0/html/common/lightning_module.html

akihironitta commented on docs/source/common/optimizers.rst:
```diff
@@ -159,32 +162,98 @@ Here is the same example as above using a ``closure``.
+Use ``self.lr_schedulers()`` in  your :class:`~pytorch_lightning.LightningModule` to access any learning rate schedulers
+defined in your :meth:`~pytorch_lightning.LightningModule.configure_optimizers`.
+
+.. warning::
+   * Before 1.3, Lightning automatically called ``lr_scheduler.step()`` in both automatic and manual optimization. From
+     1.3, ``lr_scheduler.step()`` is now for the user to call at arbitrary intervals.
```
The manual lr_scheduler.step PR (#6825) couldn't make it to 1.3? If so, this needs to be updated.

awaelchli replied:
In my opinion they should both make it in 
cc @edenlightning

Borda commented:
@edenlightning @carmocca are we set here?

tchaton reviewed (approved):
LGTM !
